### PR TITLE
Docs skeleton

### PR DESCRIPTION
**Pull request by maikia**
**note:** shifted from the PR I mistakenly  did on `substra/substra`

This is to be a skeleton for the new documentation of Substra as discussed with @RomainGoussault and @Esadruhn
The sections not filled here will be taken care of in a different PR, here:

- we set the Sphinx structure for the documentation (using .rst files) to be filled with text + references
- set the Example gallery, here we are using Sphinx-gallery (therefore existing Example directory is renamed to code_examples)
- nclude 1 Titanic example (to be filled with more comments and references)

The Theme has been chosen randomly and will be updated in the future

**IMPORTANT**
 

- [ ] Before merging in the `docs/source/conf.py` file binder branch `'branch': 'docs-skeleton'` must be changed to `master` (otherwise the binder link will still be connected to this branch and will not update to the new examples). It is set to this branch for now for testing purposes
- [ ] the files necessary for examples (eg assets) must be made downloadable, and then `code_examples` directory removed

preview of the main page:
substra-docs-screenshot

### CODE REVIEW

File: docs/Makefile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+SPHINXOPTS    ?=
+SPHINXBUILD   ?= sphinx-build
+SOURCEDIR     = source
+BUILDDIR      = build
+
+# Put it first so that "make" without argument is like "make help".
+help:
+	@$(SPHINXBUILD) -M help "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O)
+
+.PHONY: help Makefile
```
Comment (Esadruhn) [inline docs/Makefile:15]:
```suggestion
.PHONY: help livehtml docs
```

File: docs/source/conf.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+# If extensions (or modules to document with autodoc) are in another directory,
+# add these directories to sys.path here. If the directory is relative to the
+# documentation root, use os.path.abspath to make it absolute, like shown here.
+#
+import os
+import re
+import sphinx_rtd_theme
+
+# TODO: get current substra version
+def _get_version():
```
Comment (Esadruhn) [inline docs/source/conf.py:18]: Update this to get the version: will be in another PR, we need to put in place the "head version" in Substra compatibility table + do releases on this repo

You can delete this function in the meantime

**Review by Esadruhn — CHANGES_REQUESTED** [review]

File: docs/Makefile [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# Minimal makefile for Sphinx documentation
```
Comment (Esadruhn) [inline docs/Makefile:1]: Need to delete the root Makefile, otherwise it looks like we can build the doc from the root Makefile:

```
Application error:
Cannot find source directory (/Users/tdeboisfosse/projects/opensource/substra-documentation/src)
make: *** [html] Error 2
```

Comment (Esadruhn) [reply docs/Makefile:1]: + the root conf.py I guess

File: docs/source/conf.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+pygments_style = 'sphinx'
+html_theme = 'sphinx_rtd_theme'
+
+# Add any paths that contain custom themes here, relative to this directory.
+html_theme_path = [sphinx_rtd_theme.get_html_theme_path()]
+
+# Add any paths that contain custom static files (such as style sheets) here,
+# relative to this directory. They are copied after the builtin static files,
+# so a file named "default.css" will overwrite the builtin "default.css".
+html_static_path = ['_static']
```
Comment (Esadruhn) [inline docs/source/conf.py:140]:
```suggestion
html_static_path = []
```

You get a warning when building the doc: `WARNING: html_static_path entry '_static' does not exist`

File: examples/plot_titanic.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+# TODO: add a link and the instructions on how to get the assets folder (.zip??) and update the path if necessary
+#
+# TODO: in the assets directory you can find XXX files including datafiles which we will use next
+# TODO: a link to some example on how to prepare your own assets + each file explained??
+#
+
+url_assets = 'https://github.com/Substra/substra-documentation/blob/b38747fab3b7ba4cc9e95f511d0c761348ccaf65/code_examples/assets.zip?raw=true'
+http_response = urlopen(url_assets)
+zipedfile = zipfile.ZipFile(BytesIO(http_response.read()))
+zipedfile.extractall()
```
Comment (Esadruhn) [inline examples/plot_titanic.py:44]: `FileExistsError: [Errno 17] File exists: substra-documentation/examples/assets/dataset`

There is an error if we already built the docs once apparently

**Review by Esadruhn — CHANGES_REQUESTED** [review]
Great PR, thank you 🎉  ! Generating the gallery, takes a very long time for me so I did not finish the review yet, a few minor comments

File: docs/source/documentation/glossary.rst [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+Glossary
+========
+
+.. hint::
+
+    Get started with glossary: Please refer to
+    `Sphinx documentation <https://sublime-and-sphinx-guide.readthedocs.io/en/latest/glossary.html>`_ on how to make a
+    glossary in Sphinx, eg:
+
+    .. glossary::
```
Comment (Esadruhn) [inline docs/source/documentation/glossary.rst:12]: very nice ! I did not know this feature

**Comment by Esadruhn** [comment]
For the API reference, we need to:
- clone the substra repo in the CI, build the API doc and include it in the API reference paragraph
- include the generated SDK and CLI doc from Substra: https://github.com/Substra/substra/tree/master/references

Let's do this in another card / PR

File: docs/Makefile (hunk 1 above)
Comment (maikia) [reply docs/Makefile:15]: Could you please explain?

File: docs/Makefile (hunk 3 above)
Comment (maikia) [reply docs/Makefile:1]: How do you get this error?

**Comment by maikia** [comment]
as for taking long to build the docs: in normal case It should only take long the first time (when you are building the example library).
If nothing changes in examples it should not be long afterwards. (Use `make html` to build it).
If it still takes long after the first time let me know, we need to find the reason before the docs grow

**Comment by maikia** [comment]
Ps. Do we really  need sign-off? I keep forgetting about it :-/

File: examples/plot_titanic.py (hunk 5 above)
Comment (maikia) [reply examples/plot_titanic.py:44]: I am unable to reproduce it. Could you please provide some info on how you got it?

File: docs/Makefile (hunk 3 above)
Comment (Esadruhn) [reply docs/Makefile:1]: if I try to call a `make` command from the root directory
now we have to go to the `docs` directory to run make

File: docs/Makefile (hunk 1 above)
Comment (Esadruhn) [reply docs/Makefile:15]: In phony you should put the list of the commands you define in the Makefile. That way, if there is a file in the same directory with the command name as name, the command will still be executed.
Apparently there is also a performance gain

More explanations here: https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html

> The implicit rule search (see Implicit Rules) is skipped for .PHONY targets. This is why declaring a target as .PHONY is good for performance, even if you are not worried about the actual file existing.
